src/NuGet.Core/NuGet.Protocol.Core.v3/HttpSource/HttpSource.cs

```diff
@@ -0,0 +1,607 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Diagnostics;
+using System.Globalization;
+using System.IO;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using System.Security.Cryptography;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+using NuGet.Common;
+using NuGet.Configuration;
+using NuGet.Logging;
+using NuGet.Protocol.Core.Types;
+using NuGet.Protocol.Core.v3;
+
+namespace NuGet.Protocol
+{
+    public class HttpSource : IDisposable
+    {
+        private const int BufferSize = 8192;
+        private readonly Func<Task<HttpHandlerResource>> _messageHandlerFactory;
+        private readonly Uri _baseUri;
+        private HttpClient _httpClient;
+        private bool _disposed;
+        private int _authRetries;
+
+        // Only one thread may re-create the http client at a time.
+        private readonly object _httpClientLockObject = new object();
+        private SemaphoreSlim _httpClientLock;
+
+        // In order to avoid too many open files error, set concurrent requests number to 16 on Mac
+        private readonly static int ConcurrencyLimit = 16;
+
+        // Only one source may prompt at a time
+        private readonly static SemaphoreSlim _credentialPromptLock = new SemaphoreSlim(1, 1);
+
+        // Limiting concurrent requests to limit the amount of files open at a time on Mac OSX
+        // the default is 256 which is easy to hit if we don't limit concurrency
+        private readonly static SemaphoreSlim _throttle =
```

Comment:
since we log throttling information per request, we should probably log (once) throttling top level info (i.e. throttling requests up to XXX/Not throttling requests)